File: backend/app/services/transcript_service.py
"""
Service layer for transcript-related business logic.
Handles file upload, validation, storage, and database operations.
"""

import os
import uuid
import aiofiles
from pathlib import Path
from datetime import datetime
from typing import List, Optional, Tuple, Dict, Any
from fastapi import UploadFile
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError

from app.config import get_settings
from app.models.transcript import Transcript
from app.schemas.transcript import TranscriptCreate, TranscriptResponse
from app.utils.file_handler import TranscriptParser, FileValidationError
from app.utils.logger import get_logger, set_correlation_id, get_correlation_id

logger = get_logger(__name__)
settings = get_settings()


class TranscriptNotFoundError(Exception):
    """Raised when a transcript is not found in the database."""
    pass



class TranscriptService:
    """
    Service class for handling transcript operations.
    
    Provides methods for file upload, validation, storage, and retrieval
    with comprehensive logging and error handling.
    """

    def __init__(self, db: Session) -> None:
        """
        Initialize the transcript service.
        
        Args:
            db: Database session for operations
        """
        self.db: Session = db
        self.parser: TranscriptParser = TranscriptParser()

    async def upload_transcript(self, file: UploadFile) -> TranscriptResponse:
        """
        Upload and process a transcript file.
        
        Args:
            file: Uploaded file object from FastAPI
            
        Returns:
            TranscriptResponse with created transcript details
            
        Raises:
            FileValidationError: If file validation fails
            DuplicateTranscriptError: If transcript already exists
        """
        correlation_id = get_correlation_id()
        
        logger.info("File upload received", 
                   filename=file.filename, 
                   content_type=file.content_type,
                   size_bytes=file.size if hasattr(file, 'size') else 'unknown')

        # Validate file extension
        self.parser.validate_file_extension(file.filename, settings.allowed_extensions)
        
        # Read file content
        content = await file.read()
        
        # Validate file size
        self.parser.validate_file_size(len(content), settings.max_file_size)
        
        # Decode content
        try:
            content_str = content.decode('utf-8')
        except UnicodeDecodeError:
            logger.error("File encoding error", filename=file.filename)
            raise FileValidationError("File must be valid UTF-8 encoded text")

        logger.info("File validation passed", filename=file.filename)

        # Parse transcript content
        try:
            transcript_text, metadata, word_count = self.parser.parse_transcript_file(
                file.filename, content_str
            )
            
            # Calculate content hash for deduplication
            content_hash = self.parser.calculate_content_hash(transcript_text)
            
            logger.info("File parsed successfully", 
                       filename=file.filename, 
                       word_count=word_count,
                       content_hash=content_hash[:16])  # Log first 16 chars of hash

        except FileValidationError:
            raise
        except Exception as e:
            logger.error("File parsing failed", filename=file.filename, error=str(e))
            raise FileValidationError(f"Failed to parse transcript file: {str(e)}")

        # Check for duplicate content and handle overwrite
        existing_transcript = self.get_transcript_by_hash(content_hash)
        if existing_transcript:
            logger.info("Duplicate transcript detected, updating existing record", 
                       filename=file.filename,
                       existing_id=existing_transcript.id,
                       existing_filename=existing_transcript.filename)
            
            # Update existing transcript with new file information
            try:
                # Update transcript metadata and filename
                existing_transcript.filename = file.filename
                existing_transcript.word_count = word_count
                existing_transcript.transcript_metadata = metadata
                existing_transcript.uploaded_at = datetime.utcnow()
                
                self.db.commit()
                self.db.refresh(existing_transcript)
                
                logger.info("Updated existing transcript record", 
                           transcript_id=existing_transcript.id,
                           filename=existing_transcript.filename)
                
                return TranscriptResponse.from_orm(existing_transcript)
                
            except Exception as e:
                self.db.rollback()
                logger.error("Failed to update existing transcript", 
                            transcript_id=existing_transcript.id, 
                            error=str(e))
                raise

        # Generate unique filename and save file
        file_id = str(uuid.uuid4())
        file_extension = Path(file.filename).suffix
        unique_filename = f"{file_id}{file_extension}"
        file_path = os.path.join(settings.storage_path, unique_filename)

        try:
            # Ensure storage directory exists
            os.makedirs(settings.storage_path, exist_ok=True)
            
            # Save file to storage
            async with aiofiles.open(file_path, 'w', encoding='utf-8') as f:
                await f.write(transcript_text)
            
            logger.info("File saved to storage", 
                       filename=file.filename,
                       storage_path=file_path)

        except Exception as e:
            logger.error("File storage failed", filename=file.filename, error=str(e))
            raise FileValidationError(f"Failed to save file: {str(e)}")

        # Create database record
        try:
            transcript_data = TranscriptCreate(
                filename=file.filename,
                content_hash=content_hash,
                word_count=word_count,
                transcript_metadata=metadata
            )
            
            transcript = self.create_transcript(transcript_data, file_path)
            
            logger.info("Transcript record created", 
                       transcript_id=transcript.id,
                       filename=transcript.filename)

            return TranscriptResponse.from_orm(transcript)

        except Exception as e:
            # Clean up saved file if database operation fails
            try:
                os.remove(file_path)
                logger.info("Cleaned up file after database error", file_path=file_path)
            except:
                pass
            
            logger.error("Database operation failed", filename=file.filename, error=str(e))
            raise

    def create_transcript(self, transcript_data: TranscriptCreate, file_path: str) -> Transcript:
        """
        Create a new transcript record in the database.
        
        Args:
            transcript_data: Validated transcript data
            file_path: Path to the stored transcript file
            
        Returns:
            Created Transcript model instance
        """
        transcript = Transcript(
            filename=transcript_data.filename,
            file_path=file_path,
            content_hash=transcript_data.content_hash,
            word_count=transcript_data.word_count,
            transcript_metadata=transcript_data.transcript_metadata
        )
        
        try:
            self.db.add(transcript)
            self.db.commit()
            self.db.refresh(transcript)
            
            logger.info("Created transcript record",
                       transcript_id=transcript.id,
                       filename=transcript.filename,
                       word_count=transcript.word_count)
            
            return transcript
            
        except IntegrityError as e:
            self.db.rollback()
            logger.error("Database integrity error", error=str(e))
            raise

    def get_transcript_by_id(self, transcript_id: uuid.UUID) -> Optional[Transcript]:
        """
        Retrieve a transcript by its ID.
        
        Args:
            transcript_id: UUID of the transcript to retrieve
            
        Returns:
            Transcript model instance or None if not found
        """
        return self.db.query(Transcript).filter(Transcript.id == transcript_id).first()

    def get_transcript_by_hash(self, content_hash: str) -> Optional[Transcript]:
        """
        Retrieve a transcript by its content hash.
        
        Args:
            content_hash: SHA-256 hash of the transcript content
            
        Returns:
            Transcript model instance or None if not found
        """
        return self.db.query(Transcript).filter(Transcript.content_hash == content_hash).first()

    def list_transcripts(self, skip: int = 0, limit: int = 100) -> Tuple[List[Transcript], int]:
        """
        List transcripts with pagination.
        
        Args:
            skip: Number of records to skip
            limit: Maximum number of records to return
            
        Returns:
            Tuple of (transcript_list, total_count)
        """
        query = self.db.query(Transcript).order_by(Transcript.uploaded_at.desc())
        total = query.count()
        transcripts = query.offset(skip).limit(limit).all()
        
        logger.info("Listed transcripts", count=len(transcripts), total=total, skip=skip, limit=limit)
        
        return transcripts, total

    def delete_transcript(self, transcript_id: uuid.UUID) -> bool:
        """
        Delete a transcript and its associated file.
        
        Args:
            transcript_id: UUID of the transcript to delete
            
        Returns:
            True if deleted successfully, False if not found
        """
        transcript = self.get_transcript_by_id(transcript_id)
        if not transcript:
            logger.warning("Transcript not found for deletion", transcript_id=transcript_id)
            return False

        try:
            # Delete the file from storage
            if os.path.exists(transcript.file_path):
                os.remove(transcript.file_path)
                logger.info("Deleted transcript file", file_path=transcript.file_path)

            # Delete the database record
            self.db.delete(transcript)
            self.db.commit()
            
            logger.info("Deleted transcript record", 
                       transcript_id=transcript_id, 
                       filename=transcript.filename)
            
            return True

        except Exception as e:
            self.db.rollback()
            logger.error("Failed to delete transcript", 
                        transcript_id=transcript_id, 
                        error=str(e))
            raise

    async def read_transcript_content(self, transcript: Transcript) -> str:
        """
        Read the content of a transcript file.
        
        Args:
            transcript: Transcript model instance
            
        Returns:
            Content of the transcript file
            
        Raises:
            TranscriptNotFoundError: If transcript file is missing
        """
        if not os.path.exists(transcript.file_path):
            logger.error("Transcript file not found", 
                        transcript_id=transcript.id, 
                        file_path=transcript.file_path)
            raise TranscriptNotFoundError(f"Transcript file not found: {transcript.file_path}")

        try:
            async with aiofiles.open(transcript.file_path, 'r', encoding='utf-8') as f:
                content = await f.read()
            
            logger.info("Read transcript content", 
                       transcript_id=transcript.id, 
                       content_length=len(content))
            
            return content
            
        except Exception as e:
            logger.error("Failed to read transcript file", 
                        transcript_id=transcript.id, 
                        error=str(e))
            raise TranscriptNotFoundError(f"Failed to read transcript file: {str(e)}")

File: internal/services/transcript_service.go
package services

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"os"
	"path/filepath"
	"backend-golang/internal/config"
	"backend-golang/internal/models"
	"backend-golang/pkg/logger"
	"strings"
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// TranscriptServiceInterface defines the interface for transcript service operations
type TranscriptServiceInterface interface {
	UploadTranscript(req *UploadTranscriptRequest, correlationID string) (*UploadTranscriptResponse, error)
	GetTranscripts(page, perPage int) ([]*models.Transcript, int64, error)
	GetTranscript(id uuid.UUID) (*models.Transcript, error)
	DeleteTranscript(id uuid.UUID, correlationID string) error
	ReadTranscriptContent(transcript *models.Transcript) (string, error)
}

type TranscriptService struct {
	db     *gorm.DB
	config *config.Config
}

func NewTranscriptService(db *gorm.DB, cfg *config.Config) *TranscriptService {
	return &TranscriptService{
		db:     db,
		config: cfg,
	}
}

// UploadTranscriptRequest represents the upload request
type UploadTranscriptRequest struct {
	File *multipart.FileHeader
}

// UploadTranscriptResponse represents the upload response
type UploadTranscriptResponse struct {
	TranscriptID uuid.UUID `json:"transcript_id"`
	Filename     string    `json:"filename"`
	WordCount    int       `json:"word_count"`
	Message      string    `json:"message"`
}

// UploadTranscript handles file upload and validation
func (s *TranscriptService) UploadTranscript(req *UploadTranscriptRequest, correlationID string) (*UploadTranscriptResponse, error) {
	log := logger.WithCorrelationID(correlationID)

	// Validate file extension
	ext := strings.ToLower(filepath.Ext(req.File.Filename))
	isValidExt := false
	for _, allowedExt := range s.config.AllowedExts {
		if ext == allowedExt {
			isValidExt = true
			break
		}
	}
	if !isValidExt {
		return nil, fmt.Errorf("invalid file extension: %s. Allowed: %v", ext, s.config.AllowedExts)
	}

	// Validate file size
	if req.File.Size > s.config.MaxFileSize {
		return nil, fmt.Errorf("file too large: %d bytes. Maximum: %d bytes", req.File.Size, s.config.MaxFileSize)
	}

	// Open and read file
	file, err := req.File.Open()
	if err != nil {
		logger.LogErrorWithStackAndCorrelation(err, correlationID, map[string]interface{}{
			"filename":  req.File.Filename,
			"operation": "open_upload_file",
		})
		return nil, fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()

	content, err := io.ReadAll(file)
	if err != nil {
		logger.LogErrorWithStackAndCorrelation(err, correlationID, map[string]interface{}{
			"filename":  req.File.Filename,
			"operation": "read_file_content",
		})
		return nil, fmt.Errorf("failed to read file: %w", err)
	}

	// Validate UTF-8 encoding
	if !isValidUTF8(content) {
		return nil, fmt.Errorf("file must be UTF-8 encoded")
	}

	// Calculate content hash
	hash := sha256.Sum256(content)
	contentHash := hex.EncodeToString(hash[:])

	// Check for duplicates
	var existingTranscript models.Transcript
	if err := s.db.Where("content_hash = ?", contentHash).First(&existingTranscript).Error; err == nil {
		log.WithField("existing_id", existingTranscript.ID).Info("Duplicate transcript detected")
		return nil, fmt.Errorf("duplicate transcript already exists with ID: %s", existingTranscript.ID)
	}

	// Parse content and calculate word count
	wordCount, metadata, err := s.parseTranscriptContent(content, ext)
	if err != nil {
		logger.LogErrorWithStackAndCorrelation(err, correlationID, map[string]interface{}{
			"filename":   req.File.Filename,
			"extension":  ext,
			"operation":  "parse_transcript_content",
		})
		return nil, fmt.Errorf("failed to parse transcript: %w", err)
	}

	// Create transcript record
	transcript := &models.Transcript{
		ID:                 uuid.New(),
		Filename:           req.File.Filename,
		ContentHash:        contentHash,
		WordCount:          wordCount,
		TranscriptMetadata: metadata,
		UploadedAt:         time.Now(),
	}

	// Save file to storage
	filePath, err := s.saveFile(transcript.ID, content)
	if err != nil {
		logger.LogErrorWithStackAndCorrelation(err, correlationID, map[string]interface{}{
			"transcript_id": transcript.ID,
			"filename":      req.File.Filename,
			"operation":     "save_file",
		})
		return nil, fmt.Errorf("failed to save file: %w", err)
	}
	transcript.FilePath = filePath

	// Save to database
	if err := s.db.Create(transcript).Error; err != nil {
		// Clean up file if database save fails
		_ = os.Remove(filePath)
		logger.LogErrorWithStackAndCorrelation(err, correlationID, map[string]interface{}{
			"transcript_id": transcript.ID,
			"filename":      req.File.Filename,
			"operation":     "save_transcript_to_database",
		})
		return nil, fmt.Errorf("failed to save transcript to database: %w", err)
	}

	log.WithFields(map[string]interface{}{
		"transcript_id": transcript.ID,
		"filename":      transcript.Filename,
		"word_count":    transcript.WordCount,
		"file_size":     req.File.Size,
	}).Info("Transcript uploaded successfully")

	return &UploadTranscriptResponse{
		TranscriptID: transcript.ID,
		Filename:     transcript.Filename,
		WordCount:    transcript.WordCount,
		Message:      "Transcript uploaded successfully",
	}, nil
}

// GetTranscripts returns paginated list of transcripts
func (s *TranscriptService) GetTranscripts(page, perPage int) ([]*models.Transcript, int64, error) {
	var transcripts []*models.Transcript
	var total int64

	offset := (page - 1) * perPage

	// Count total
	if err := s.db.Model(&models.Transcript{}).Count(&total).Error; err != nil {
		logger.LogErrorWithStack(err, map[string]interface{}{
			"operation": "count_transcripts",
			"page":      page,
			"per_page":  perPage,
		})
		return nil, 0, fmt.Errorf("failed to count transcripts: %w", err)
	}

	// Get paginated results
	if err := s.db.Offset(offset).Limit(perPage).Order("uploaded_at DESC").Find(&transcripts).Error; err != nil {
		logger.LogErrorWithStack(err, map[string]interface{}{
			"operation": "get_transcripts_list",
			"page":      page,
			"per_page":  perPage,
			"offset":    offset,
		})
		return nil, 0, fmt.Errorf("failed to get transcripts: %w", err)
	}

	return transcripts, total, nil
}

// GetTranscript returns a single transcript by ID
func (s *TranscriptService) GetTranscript(id uuid.UUID) (*models.Transcript, error) {
	var transcript models.Transcript
	if err := s.db.Where("id = ?", id).First(&transcript).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, fmt.Errorf("transcript not found")
		}
		logger.LogErrorWithStack(err, map[string]interface{}{
			"transcript_id": id,
			"operation":     "get_transcript_by_id",
		})
		return nil, fmt.Errorf("failed to get transcript: %w", err)
	}
	return &transcript, nil
}

// DeleteTranscript deletes a transcript and its file
func (s *TranscriptService) DeleteTranscript(id uuid.UUID, correlationID string) error {
	log := logger.WithCorrelationID(correlationID)

	var transcript models.Transcript
	if err := s.db.Where("id = ?", id).First(&transcript).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return fmt.Errorf("transcript not found")
		}
		logger.LogErrorWithStackAndCorrelation(err, correlationID, map[string]interface{}{
			"transcript_id": id,
			"operation":     "find_transcript_for_delete",
		})
		return fmt.Errorf("failed to find transcript: %w", err)
	}

	// Delete file
	if err := os.Remove(transcript.FilePath); err != nil && !os.IsNotExist(err) {
		log.WithError(err).Warn("Failed to delete transcript file")
	}

	// Delete from database (cascade deletes analyses and fact checks)
	if err := s.db.Delete(&transcript).Error; err != nil {
		logger.LogErrorWithStackAndCorrelation(err, correlationID, map[string]interface{}{
			"transcript_id": id,
			"operation":     "delete_transcript_from_database",
		})
		return fmt.Errorf("failed to delete transcript from database: %w", err)
	}

	log.WithField("transcript_id", id).Info("Transcript deleted successfully")
	return nil
}

// Helper functions

func (s *TranscriptService) saveFile(transcriptID uuid.UUID, content []byte) (string, error) {
	// Ensure storage directory exists
	if err := os.MkdirAll(s.config.StoragePath, 0755); err != nil {
		logger.LogErrorWithStack(err, map[string]interface{}{
			"storage_path": s.config.StoragePath,
			"operation":    "create_storage_directory",
		})
		return "", fmt.Errorf("failed to create storage directory: %w", err)
	}

	filename := transcriptID.String() + ".txt"
	filePath := filepath.Join(s.config.StoragePath, filename)

	if err := os.WriteFile(filePath, content, 0644); err != nil {
		logger.LogErrorWithStack(err, map[string]interface{}{
			"file_path":     filePath,
			"transcript_id": transcriptID,
			"operation":     "write_file",
		})
		return "", fmt.Errorf("failed to write file: %w", err)
	}

	return filePath, nil
}

func (s *TranscriptService) parseTranscriptContent(content []byte, ext string) (int, []byte, error) {
	var wordCount int
	var metadata map[string]interface{}

	if ext == ".json" {
		var jsonData map[string]interface{}
		if err := json.Unmarshal(content, &jsonData); err != nil {
			logger.LogErrorWithStack(err, map[string]interface{}{
				"operation": "unmarshal_json_transcript",
			})
			return 0, nil, fmt.Errorf("invalid JSON format: %w", err)
		}

		// Extract metadata
		metadata = make(map[string]interface{})
		for key, value := range jsonData {
			if key != "transcript" {
				metadata[key] = value
			}
		}

		// Count words in transcript array or text field
		if transcript, ok := jsonData["transcript"]; ok {
			if transcriptArray, ok := transcript.([]interface{}); ok {
				// Array format: [{"text": "...", "speaker": "...", "timestamp": "..."}, ...]
				for _, item := range transcriptArray {
					if itemMap, ok := item.(map[string]interface{}); ok {
						if text, ok := itemMap["text"].(string); ok {
							wordCount += countWords(text)
						}
					}
				}
			} else if transcriptText, ok := transcript.(string); ok {
				// String format
				wordCount = countWords(transcriptText)
			}
		}
	} else {
		// Plain text format
		wordCount = countWords(string(content))
	}

	metadataBytes, _ := json.Marshal(metadata)
	return wordCount, metadataBytes, nil
}

func countWords(text string) int {
	words := strings.Fields(strings.TrimSpace(text))
	return len(words)
}

// ReadTranscriptContent reads the content of a transcript file (matches Python async def read_transcript_content)
func (s *TranscriptService) ReadTranscriptContent(transcript *models.Transcript) (string, error) {
	if _, err := os.Stat(transcript.FilePath); os.IsNotExist(err) {
		logger.Log.WithFields(map[string]interface{}{
			"transcript_id": transcript.ID,
			"file_path": transcript.FilePath,
		}).Error("Transcript file not found")
		return "", fmt.Errorf("transcript file not found: %s", transcript.FilePath)
	}

	content, err := os.ReadFile(transcript.FilePath)
	if err != nil {
		logger.LogErrorWithStack(err, map[string]interface{}{
			"transcript_id": transcript.ID,
			"file_path":     transcript.FilePath,
			"operation":     "read_transcript_file",
		})
		return "", fmt.Errorf("failed to read transcript file: %w", err)
	}

	logger.Log.WithFields(map[string]interface{}{
		"transcript_id": transcript.ID,
		"content_length": len(content),
	}).Info("Read transcript content")

	return string(content), nil
}

func isValidUTF8(data []byte) bool {
	return strings.ToValidUTF8(string(data), "") == string(data)
}